src/acl/SplayInserter.h

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright (C) 1996-2023 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#ifndef SQUID_SRC_ACL_SPLAYINSERTER_H
+#define SQUID_SRC_ACL_SPLAYINSERTER_H
+
+#include "acl/Acl.h"
+#include "debug/Stream.h"
+#include "globals.h"
+#include "splay.h"
+
+namespace Acl {
+
+/// Helps populate a Splay tree with configured ACL parameter values and their
+/// duplicate-handling derivatives (each represented by a DataValue object).
+template <class DataValue>
+class SplayInserter
+{
+public:
+    using Value = DataValue;
+
+    /// If necessary, updates the splay tree to match all individual values that
+    /// match the given parsed ACL parameter value. If the given value itself is
+    /// not added to the tree (e.g., because it is a duplicate), it is destroyed
+    /// using DestroyValue(). Otherwise, the given value will be destroyed
+    /// later, during subsequent calls to this method or free_acl().
+    static void Merge(Splay<Value> &, Value &&);
+
+private:
+    /// SplayInserter users are expected to specialize all or most of the static
+    /// methods below. Most of these methods have no generic implementation.
+
+    /// A Splay::SPLAYCMP function for comparing parsed ACL parameter values.
+    /// This function must work correctly with all valid ACL parameter values,
+    /// including those representing sets or ranges. The order specified by this
+    /// function must be the same as the order specified by the SPLAYCMP
+    /// function used later by ACL::match().
+    /// \retval 0 The two values overlap. Here, two values overlap if they are
+    /// identical, if one contains all values from another, and if one contains
+    /// at least one value from another.
+    static int Compare(const Value &, const Value &);
+
+    /// whether the set of values matched by `a` contains the entire set of
+    /// values matched by `b`, including cases where `a` is identical to `b`
+    /// \prec The two values overlap: Compare(a, b) == 0
+    static bool AcontainsEntireB(const Value &a, const Value &b);
```

Comment:
I do not like the name of this function, and it violates Squid naming style, but I cannot think of a better variant.

Reply:
Can we simply name it 'IsSuperset()', meaning that a is a superset of b? Also we could have a paired IsSubset(a, b) and use it in SplayInserter<DataValue>::Merge() instead of (a little confusing) AcontainsEntireB() with reversed arguments.

Reply:
Switched to IsSubset() in d8bb057 under the assumption that my fear of folks mismatching which subset() argument is a subset is unfounded.